devtools/sendit/stream.go

```diff
@@ -0,0 +1,189 @@
+package sendit
+
+import (
+	"bufio"
+	"errors"
+	"io"
+	"sync"
+)
+
+// Stream is a ReadWriteCloser that can have it's read/write pipe replaced safely while actively sending data.
+type Stream struct {
+	readCh    chan io.ReadCloser
+	readEOFCh chan io.ReadCloser
+	writeCh   chan io.WriteCloser
+	closeCh   chan struct{}
+	mx        sync.Mutex
+
+	readyCh chan struct{}
+
+	isClosed bool
+	isReady  bool
+}
+
+// NewStream initializes a new stream. SetPipe must be called before data can be transferred.
+func NewStream() *Stream {
+	s := &Stream{
+		readCh:    make(chan io.ReadCloser, 1),
+		readEOFCh: make(chan io.ReadCloser, 1),
+		writeCh:   make(chan io.WriteCloser, 1),
+		closeCh:   make(chan struct{}),
+		readyCh:   make(chan struct{}),
+	}
+	return s
+}
+
+// Ready will return a channel indicating when the stream is ready for reading and writing.
+func (s *Stream) Ready() <-chan struct{} { return s.readyCh }
+
+func (s *Stream) Write(p []byte) (int, error) {
```

Comment:
should we check if the stream is closed before we write?
```
if s.isClosed {
	return io.ErrClosedPipe
}
```

Reply:
We are already doing so by way of `<-s.closeCh`. It uses the channel to make the check to keep it thread-safe as we don't want to access `s.isClosed` while `s.isClosed` is being set by another thread.